File: 1046-last-stone-weight/1046-last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        s = [-s for s in stones]
        
        heapq.heapify(s)

        while (len(s) > 1):
            x = heapq.heappop(s)
            y = heapq.heappop(s)
            if (y > x):
                heapq.heappush(s, x - y)
        
        s.append(0)
            
        return abs(s[0])

File: 1046-last-stone-weight/1046-last-stone-weight.cs
public class Solution {
    public int LastStoneWeight(int[] stones)
    {
        var pq = new PriorityQueue<int, int>();

        foreach (int stone in stones)
        {
            pq.Enqueue(stone, -stone); //maxheap
        }

        while (pq.Count > 1)
        {
            int x = pq.Dequeue();
            int y = pq.Dequeue();
            int diff = Math.Abs(x - y);
            if (diff > 0) pq.Enqueue(diff, -diff);
        }

        return pq.Count > 0 ? pq.Peek() : 0; 
    }
}
